### Homework 03 Paritosh Rai

#### 1.	Compare your given name with your nickname (if you don’t have a nickname, invent one for this assignment) by answering the following questions:
#### a.	What is the edit distance between your nickname and your given name?
#### b.	What is the percentage string match between your nickname and your given name?
#### Show your work for both calculations.


#####  Given Name: Paritosh

##### Nick Name: Pintu

Length of given name Paritosh is 8 and lenght of my nick name Pintu is 5

First letter is common i.e. "P" so no change rquired.


#### 1st Method using elif

In [50]:
def edit_distance(string1, string2):
    """Ref: https://bit.ly/2Pf4a6Z"""

    if len(string1) > len(string2):
        difference = len(string1) - len(string2)
        str = string2[:]

    elif len(string2) > len(string1):
        difference = len(string2) - len(string1)
        str = string1[:]

    else:
        difference = 0
        str = string1[:]

    
    for i in range(len(str)-1):
        if string1[i] != string2[i]:
            difference += 1

    return difference

In [51]:
print("Edit Distance between my Nick name Pintu and my given name Paritosh is "+ str(edit_distance("Pintu","Paritosh" )))
 

Edit Distance between my Nick name Pintu and my given name Paritosh is 6


#### 2nd Method Using nltk

In [52]:
import nltk
# string decleration
source = 'Pintu'
target = 'Paritosh'
#distance calculation
edit_distance_nltk=nltk.edit_distance(source , target)
edit_distance_nltk
print("Edit Distance between my Nick name Pintu and my given name Paritosh is " + str(edit_distance_nltk))

Edit Distance between my Nick name Pintu and my given name Paritosh is 6


#### b. What is the percentage string match between your nickname and your given name?
#### Show your work for both calculations.

In [53]:
"""https://www.geeksforgeeks.org/python-count-the-number-of-matching-characters-in-a-pair-of-string/"""

def count(str1 ,str2) :
    # store unique letters from the string1
    #
    set_string1 = set(str1)
  
    # store unique letters from the string1
    set_string2 = set(str2)
  
    # using (&) intersection mathematical operation on sets
   
    matched_characters = set_string1 & set_string2
    
    # double the number of match characters and devide by sum of length of both names
    percent_match = 100 * 2*(len(matched_characters))/(len(str1) + len(str2))
      
    # printing the length of matched_characters set
    print("No. of matching characters are : " + str(len(matched_characters)) )
    print("Total number of characters in Given_Name + Nick_Name : " ,(len(str1) + len(str2)))
    print("Percentage of matching characters are : " + str(percent_match) + " %") # change float to string
  
  

In [54]:
no_of_match = count("Pintu", "Paritosh")

No. of matching characters are : 3
Total number of characters in Given_Name + Nick_Name :  13
Percentage of matching characters are : 46.15384615384615 %


#### 2.	Find a friend (or family member or classmate) who you know has read a certain book. Without your friend knowing, copy the first two sentences of that book. Now rewrite the words from those sentences, excluding stop words. Now tell your friend to guess which book the words are from by reading them just that list of words. Did you friend correctly guess the book on the first try? What did he or she guess? Explain why you think you friend either was or was not able to guess the book from hearing the list of words. 

I picked "Harry Potter and the Sorcerer's Stone" and picked few sentences from the book. I asked my daughter to guess the book after removing the stop words. 

In [55]:
'''https://stackabuse.com/removing-stop-words-from-strings-in-python/#usingpythonsnltklibrary'''


from nltk.corpus import stopwords # import stop words form NLTK
nltk.download('stopwords') # download stop words
from nltk.tokenize import word_tokenize # import tokenizer

part_book= "Mr. Dursley stood rooted to the spot. He had been hugged by a complete stranger. He also thought he had been called a Muggle, whatever that was. He was rattled. He hurried to his car and set off for home, hoping he was imagining things, which he had never hoped before, because he didn't approve of imagination"
text_tokens = word_tokenize(part_book)
#remove stop words from the sentence
tokens_without_sw = [word for word in text_tokens if not word in stopwords.words()] 
print(tokens_without_sw)

[nltk_data] Downloading package stopwords to
[nltk_data]     C:\Users\parit\AppData\Roaming\nltk_data...
[nltk_data]   Package stopwords is already up-to-date!


['Mr.', 'Dursley', 'stood', 'rooted', 'spot', '.', 'He', 'hugged', 'complete', 'stranger', '.', 'He', 'thought', 'called', 'Muggle', ',', 'whatever', '.', 'He', 'rattled', '.', 'He', 'hurried', 'car', 'set', 'home', ',', 'hoping', 'imagining', 'things', ',', 'never', 'hoped', ',', "n't", 'approve', 'imagination']


**Which book the words are from?**

She guessed the book correctly on the first try, "Harry Potter." 

**What did he or she guess?**
Then I asked her which book of Harry Potter? She respounded, "Harry Potter and the Sorcerer's Stone."  That was the correct answer. 

**Explain why you think you friend either was or was not able to guess the book from hearing the list of words.**

I asked her how she guessed?  She said from Mr. Dursley (uncle of Harry Potter) and the word Muggle (word used for human in the book). That was aligned with my guess. 

I further asked her, but these might be appearing in all four or five books? She said Mr. Dursley appears in 1st two books only after that there is no mention, and she kind of guessed it is 1st one. 


#### 3.	Run one of the stemmers available in Python. Run the same two sentences from question 2 above through the stemmer and show the results. How many of the outputted stems are valid morphological roots of the corresponding words? Express this answer as a percentage.

#### Method 1 : With Tokenizer with Punctuation

In [56]:
'''https://www.datacamp.com/community/tutorials/stemming-lemmatization-python'''

import nltk
ps = nltk.stem.PorterStemmer()

from nltk.tokenize import sent_tokenize, word_tokenize # tokenize word and sentence

#stem each word in the sentence and return a combined sentence.
# define a function to stem 
def stemSentence(sentence): 
    token_words=word_tokenize(sentence) # tokenize each word
    token_words
    stem_sentence=[]
    #break each word of sentence,stem and add to sentence
    for word in token_words:
        stem_sentence.append(ps.stem(word)) # stem each word
        stem_sentence.append(" ") # add stem words to sentence
    return "".join(stem_sentence) # create sentence



In [57]:
x=stemSentence("Mr. Dursley stood rooted to the spot. He had been hugged by a complete stranger. He also thought he had been called a Muggle, whatever that was. He was rattled. He hurried to his car and set off for home, hoping he was imagining things, which he had never hoped before, because he didn't approve of imagination")

print(x)

mr. dursley stood root to the spot . he had been hug by a complet stranger . he also thought he had been call a muggl , whatev that wa . he wa rattl . he hurri to hi car and set off for home , hope he wa imagin thing , which he had never hope befor , becaus he did n't approv of imagin 


In [58]:
# count total number of words in the string
# initialize string

# default separator: space
no_words_in_text = len(x.split())

print("There are " + str(no_words_in_text) + " words.")

There are 66 words.


 #### Valid morphological roots 

In [59]:
# create a function with nltk corpus 
import nltk
def valid_word(sent):
    words = set(nltk.corpus.words.words()) 
    return " ".join(w for w in nltk.wordpunct_tokenize(sent) \

         if w.lower() in words or not w.isalpha())



In [60]:
valid_text= valid_word(valid_word(part_book)) # run with text of book
no_of_valid_text = len(valid_text.split()) # count number of token with punchuation
print(valid_text)

. stood rooted to the spot . He had been by a complete stranger . He also thought he had been a , whatever that was . He was rattled . He hurried to his car and set off for home , he was , which he had never hoped before , because he ' t approve of imagination


In [61]:
print("There are valid morphological roots " + str(no_of_valid_text) + " words.")

There are valid morphological roots 59 words.


#### Percentage of Valid Morpholigical roots

In [62]:
prct_valid_morph =100 *(no_of_valid_text)/(no_words_in_text) #calculate percentage by number of valid text and no of total words
print("Percent of valid morphological roots is " + str(prct_valid_morph) + " %.")


Percent of valid morphological roots is 89.39393939393939 %.


The total number of words (including punctuation) with PorterStemmer is 66, and the number of words (including punctuation) with valid Morphological is 59. The percentage of the valid morphological root is 89%. This is an indicator that PorterStemmer is very effective in this case.

 #### Method 2 without Punctuation

In [63]:
import nltk
ps = nltk.stem.PorterStemmer()

from nltk.tokenize import sent_tokenize, word_tokenize # tokenize word and sentence
from nltk.tokenize import RegexpTokenizer # remove punchuation

#stem each word in the sentence and return a combined sentence.
# define a function to stem without punchuation
def stemSentence_wop(sentence):
    tokenizer = RegexpTokenizer(r'\w+') # remove punctuation or filter only word
    token_words=tokenizer.tokenize(sentence) # tokenize each word remove punctuation
    token_words
    stem_sentence=[]
    #break each word of sentence,stem and add to sentence
    for word in token_words:
        stem_sentence.append(ps.stem(word)) # stem each word using Porter Stemmer
        stem_sentence.append(" ") # add stem words to sentence
    return "".join(stem_sentence) # create sentence


In [72]:
# print the text after applying Porter Stemmer 
x_wop=stemSentence_wop("Mr. Dursley stood rooted to the spot. He had been hugged by a complete stranger. He also thought he had been called a Muggle, whatever that was. He was rattled. He hurried to his car and set off for home, hoping he was imagining things, which he had never hoped before, because he didn't approve of imagination")

print(x_wop)

mr dursley stood root to the spot he had been hug by a complet stranger he also thought he had been call a muggl whatev that wa he wa rattl he hurri to hi car and set off for home hope he wa imagin thing which he had never hope befor becaus he didn t approv of imagin 


In [73]:
# count total number of words in the string
# initialize string

# default separator: space
no_words_in_text_wop = len(x_wop.split()) # count the number of words/token

print("There are " + str(no_words_in_text_wop) + " words (tokens) without punchuation.")

There are 58 words (tokens) without punchuation.


#### Valid morphological roots without Punchuation

In [74]:
# create function to stem using nltk corpus and remove punchuation
import nltk
def valid_word_wop(sent):
    tokenizer = RegexpTokenizer(r'\w+') # remove punchuation and all only word
    words = set(nltk.corpus.words.words()) # pick words only from nltk corpus
    return " ".join(w for w in tokenizer.tokenize(sent) \

         if w.lower() in words or not w.isalpha())


In [75]:
valid_text_wop= valid_word(valid_word_wop(part_book)) # run the function to print 
no_of_valid_text_wop = len(valid_text_wop.split()) #count the length of stem after removing punchuation and keeping valid words
print(valid_text_wop)

stood rooted to the spot He had been by a complete stranger He also thought he had been a whatever that was He was rattled He hurried to his car and set off for home he was which he had never hoped before because he t approve of imagination


In [76]:
print("There are valid morphological roots without punchuation " + str(no_of_valid_text_wop) + " words.")

There are valid morphological roots without punchuation 49 words.


#### Percentage of Valid Morpholigical roots without Punchuation

In [77]:
#calculate the percentage of valid stem words without puction and stem words usinger Porter Stemmer
prct_valid_morph_wop =100 *(no_of_valid_text_wop)/(no_words_in_text_wop)
print("Percent of valid morphological roots is " + str(prct_valid_morph_wop) + " %.")


Percent of valid morphological roots is 84.48275862068965 %.


Valid morphological without punctuation is little lower (84%) compared to with punctuation (89%). This is expected outcome. Depending on what we are trying do with text stemming, both have their utility. In normally validated text without punctuation make more sense to me. 